---
title: ""
output:
  html_document: default
  pdf_document: default
bibliography: ref.bib 
---
# Description of pipeline used for generation of correlation plots using ChIP-chip data (with processed expression level data)

## Introduction

This report will provide an overview of each step in the pipeline used to analyse the Upf1 ChIP-chip data and correlate Upf1 signal with indicators of gene expression (a downloaded processed RNA-seq dataset and a downloaded Pol2 ChIP-chip dataset) as well as how to produce metagene plots @de2021genome, @edwards2021genome. This includes the scripts corresponding to each step and how to use them.

## Extracting information from .CEL files

Initially, we extract the raw array data from cel files and transform it into a human-readable format using the "probe_information_gatherer.m" Matlab script. This script uses the information provided by the cdf file "Sp20b_M_v04.CDF" on probe sets and their coordinates in the array to extract this information from the raw cel files and store it a human-readable csv file titled "ProbeDataTable.csv". The cdf file corresponds to the S pombe -specific Affymetrix arrays used in the ChIP-chip experiments here and can be downloaded from the supplementary files section of https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL10187

The script can be run from within the matlab environment and could also be easily modified to call a different cdf by changing line 25 as appropriate. It requires a file in the same directory containing the paths from the script to each cel file (one path per line), which should be named "celFilePaths.txt".

## Background noise correction

The next step is to subtract the background noise of each array from the data, which is carried out as described in the "Statistical Algorithms Description Document" (http://tools.thermofisher.com/content/sfs/brochures/sadd_whitepaper.pdf), using the "noiseCorrection.R" R script. The script calls the "celfilePaths.txt" file and also the "ProbeDataTable.csv" file generated by the "probe_information_gatherer.m" script. It will generate an "adjustedIntensities.Rdata" data file and can be executed with the following shell command.

```{r eval=FALSE, echo=TRUE, warning=FALSE, message=FALSE}
Rcsript noiseCorrection.R
```


## Expression value calculation

Having subtracted the background noise, the next step is to calculate gene-specific expression values (signal values). This is done using the "expressionValueCalculation.R" script, again according to the "Statistical Algorithms Description Document" of Affymetrix. This script will call the "adjustedIntensities.Rdata" file and generate a "signalLogValues.Rdata" data file. If can be executed with the following shell command.

```{r eval=FALSE, echo=TRUE, warning=FALSE, message=FALSE}
Rcsript expressionValueCalculation.R
```


## Correlation analysis

With all data processing steps completed, the correlation analysis (as described in the Materials and Methods section of the paper) can be carried out with the "correlationPlotter.R" script. As stated, this script integrates own our Upf1 ChIP-chip data with the downloaded Pol2 ChIP-chip and RNA-seq datasets. The script calls the "celFilePaths.txt" file, the "signalLogValues.Rdata" data file, and the RNA-seq data files ("Rep1.txt" and "Rep1.txt"). It can be executed with the following shell command.

```{r eval=FALSE, echo=TRUE, warning=FALSE, message=FALSE}
Rcsript correlationPlotter.R
```

This will save the four correlation plots shown in figure 2A and 2B and will also generate a "processedDataTable.txt" file, which is a tab-delimited matrix file containing all the normalised values corresponding to those displayed in the correlation plots. While the previously mentioned scripts are ready to be applied to other datasets with minimal effort to generate gene specific expression values from raw cel files, this final script is specific to our analysis although it could still be adapted to other datasets.

## Metagene analysis

In order to produce metagene plots from the ChIP-chip data as outlined in the methods section one must find, for each probe, its position within its corresponding gene. This is carried out via the "probe_gene_information.R" script, and requires the "ProbeDataTable.csv" and "Sp20b_M_v04.CDF" files to be within the same directory, as well as the directory containing the fasta file of each gene's DNA sequence including coding sequence, introns and UTRs (https://www.pombase.org/downloads/genome-datasets). This script will string match the probe sequences in the .cdf file with the appropriate gene sequence in the fasta file to find the locatio of the probe within the gene and then associate this information with the other information contained within the .csv file corresponding to that probe. It will output a file called  "ProbeGeneInformation.csv" and can be executed with the following command.

```{r eval=FALSE, echo=TRUE, warning=FALSE, message=FALSE}
Rcsript probe_gene_information.R
```

The "makeMetagenes.R" script may then be used to generate the metagene plots described in the methodss section and shown in the supplementary information. This script requires only the "ProbeGeneInformation.csv" file. As with the correlation plotting script, this final script is specific to our analysis but could be adapted to other datasets, whereas the "probe_gene_information.R" script is more generally applicable.


## References
